### ISSUE TITLE
Switching Audio Devices Frequently Leads To Audio Output From Multiple Audio Devices

### ISSUE THREAD

**Issue by jwbats** [Belphemur/SoundSwitch#356]
Very frequently, making use of SoundSwitch's hotkey to switch between 2 audio devices, I will get output for both audio devices, rather than from the one that I want it from.

Not at the same time, mind you. One program will output to device A, while another program will output to device B.

Example:

I might be watching a video in VLC and it gives me output to my speakers, which make use of my motherboard's integrated audio chip. Now I put in my earbuds, which make use of my external soundblaster usb device.

I switch audio devices and the audio output from VLC will successfully go from my speakers to my earbuds. But now, when I play a video on Chrome, it will output to my speakers instead.

I now have 2 audio devices in use, apparently.

When this phenomenon occurs, my volume adjustments also won't be picked up on. Probably because it adjusts audio for one device, but not the other (all devices have their own volume settings in Win10).

I can usually fix this by switching audio devices a couple more times, both through SoundSwitch as well as the regular Win10 interface.

This inconsistent behavior makes SoundSwitch impossible to use for me. I had to uninstall it. Would love to see it fixed, though. I'd reinstall in a jiffy.

**Comment by FireEmerald** [comment]
> I switch audio devices and the audio output from VLC will successfully go from my speakers to my earbuds. But now, when I play a video on Chrome, it will output to my speakers instead.

Does this also behaves the same if you switch the output device using windows options? If so, this is not a issue related to SoundSwitch.

> I now have 2 audio devices in use, apparently.

That's a normal behavior. Happens if either one app does recognize the switched output device, while a second app doesn't **or** you hard-link your software output to a specific device.

**Comment by jwbats** [comment]
The problem only occurs when I use SoundSwitch, never when I switch manually via the Win10 audio device systray icon.

After uninstalling SoundSwitch, I was able to fix my problems by going into the Win10 settings for audio devices by programs.

I have a feeling SoundSwitch attempts to switch all programs to the other audio device. In my opinion, it would work more robustly if it attempted to make the other audio device the default one.

That way, the audio-device-by-program settings could just stay set to 'default audio device' and they would never have to be changed at all, thereby completely circumventing the problem.

**Comment by Belphemur** [comment]
That might be linked to the "Switch Foreground app" feature.

You can disable it in the Settings menu.

It's using the new Windows 10 feature where you can set up a Sound Device per app. What SoundSwitch does is set the switched device in Windows 10 for the currently in use application. 

When you switch again, it supposed to reset the device of all the app which in your case didn't happen. 

This feature is made to try to force application that don't switch audio to switch audio hoping that the Windows 10 change the way the audio devices works.

**Comment by jwbats** [comment]
I never used the Switch Foreground App feature.

Resetting the audio device settings for all upon 'switching back' sounds like it opens up the door to a world of confusion. Because how are you supposed to remember what your starting device was (this can differ based on settings I manually make)? That's the one you'd have to 'switch back' to in order to get all settings back to normal, presumably.

This feature must also not work as intended, seeing as how switching back and forth multiple times does not fix my issue of having Program A output on the speakers and Program B output on the earbuds.

In other words, this feature does not reset the settings for all apps. If it did, I would have all audio from a single device after multiple SoundSwitch switches. So I'm thinking there might be a bug there.

Are you telling me that turning this feature off will make SoundSwitch make the switch for all programs in Win10's settings?

That sounds way simpler to me.

**Comment by jwbats** [comment]
I just reinstalled, turned off the 'foreground' setting and turned on the 'switch default device' setting.

Not sure why the former would even have to be on when the latter already is. The latter switches everything, anyway.

Ah, well... I hate the 'foreground' setting. I'll try it like this for a few days and let you know how it went.

**Comment by jwbats** [comment]
After having a go at it with the latest version and patiently experimenting with it... still the same awkward side effects as described above.

A neat idea, this program. But it won't work well for me. I'm moving on without it for now.

Maybe I'll give it another shot in a couple of years.
